src/PostgREST/DbStructure.hs

```diff
@@ -100,74 +101,94 @@ decodeSynonyms cols =
+                  <*> HD.value HD.char)
+              <*> (parseVolatility <$> HD.value HD.char)
+
+    addName :: ProcDescription -> (Text, ProcDescription)
+    addName pd = (pdName pd, pd)
+
+    parseArgs :: Text -> [PgArg]
+    parseArgs = mapMaybe (parseArg . strip) . split (==',')
+
+    parseArg :: Text -> Maybe PgArg
+    parseArg a =
+      let (body, def) = breakOn " DEFAULT " a
+          (name, typ) = breakOn " " body in
+      if T.null typ
+         then Nothing
+         else Just $
+           PgArg (dropAround (== '"') name) (strip typ) (T.null def)
+
+    parseRetType :: Text -> Text -> Bool -> Char -> RetType
+    parseRetType schema name isSetOf typ
+      | isSetOf   = SetOf pgType
+      | otherwise = Single pgType
+      where
+        qi = QualifiedIdentifier schema name
+        pgType = case typ of
+          'c' -> Composite qi
+          'p' -> if name == "record" -- Only pg pseudo type that is a row type is 'record'
+                   then Composite qi
+                   else Scalar qi
+          _   -> Scalar qi -- 'b'ase, 'd'omain, 'e'num, 'r'ange
+
+    parseVolatility :: Char -> ProcVolatility
+    parseVolatility v | v == 'i' = Immutable
+                      | v == 's' = Stable
+                      | otherwise = Volatile -- only 'v' can happen here
+
+allProcs :: H.Query Schema (M.HashMap Text ProcDescription)
+allProcs = H.statement sql (HE.value HE.text) decodeProcs True
+  where
+    sql = [q|
+    SELECT p.proname as "proc_name",
+           d.description as "proc_description",
+           pg_get_function_arguments(p.oid) as "args",
+           tn.nspname as "rettype_schema",
+           coalesce(comp.relname, t.typname) as "rettype_name",
+           p.proretset as "rettype_is_setof",
+           t.typtype as "rettype_typ",
+           p.provolatile
+    FROM pg_proc p
+      JOIN pg_namespace pn ON pn.oid = p.pronamespace
+      JOIN pg_type t ON t.oid = p.prorettype
+      JOIN pg_namespace tn ON tn.oid = t.typnamespace
+      LEFT JOIN pg_class comp ON comp.oid = t.typrelid
+      LEFT JOIN pg_catalog.pg_description as d on d.objoid = p.oid
+    WHERE  pn.nspname = $1|]
 
-  sql = [q|
-  SELECT p.proname as "proc_name",
-         d.description as "proc_description",
-         pg_get_function_arguments(p.oid) as "args",
-         tn.nspname as "rettype_schema",
-         coalesce(comp.relname, t.typname) as "rettype_name",
-         p.proretset as "rettype_is_setof",
-         t.typtype as "rettype_typ",
-         p.provolatile
-  FROM pg_proc p
-    JOIN pg_namespace pn ON pn.oid = p.pronamespace
-    JOIN pg_type t ON t.oid = p.prorettype
-    JOIN pg_namespace tn ON tn.oid = t.typnamespace
-    LEFT JOIN pg_class comp ON comp.oid = t.typrelid
-    LEFT JOIN pg_catalog.pg_description as d on d.objoid = p.oid
-  WHERE  pn.nspname = $1|]
+accessibleProcs :: H.Query Schema (M.HashMap Text ProcDescription)
+accessibleProcs = H.statement sql (HE.value HE.text) decodeProcs True
+  where
+    sql = [q|
```

Comment:
Maybe `allProcs` and `accessibleProcs` could share a chunk of sql? The latter seems to have the same query except with an extra WHERE clause added. It could concatenate the extra clauses to the shared sql?